Add `now dev`

amio opened the pull request:
This PR add `now dev` subcommand to now-cli.

### How to try it out

1. Download the latest build of now-cli from the `now-dev` branch: https://latest-now-cli.now.sh/download?branch=now-dev
1. Mark the file as executable: `chmod +x ~/Downloads/now-macos` (adjust the path / filename if it’s different on your system)
1. From your project directory, run `~/Downloads/now-macos dev` to start the server on port 3000
   * Upon boot up, the server does an initial build of all the builds defined in `now.json`
   * After you have made some changes locally, you can rebuild an endpoint by doing a *Hard Refresh* (Cmd + Shift + R on macOS)
   * Add `--debug` to get now-cli's debugging logs
1. Comment on this PR with feedback on your experience 😃 

### How it works

Running `now dev <dir>` will:

1. start a dev server listening on 3000
2. init first build (install builders & run, for speed up further process)
3. on any request to localhost:3000
   - if there's **no** `now.json` in working dir, serve as statics
   - if `now.json`'s version !== 2, throws
   - else, serve as now v2 app:
     - lookup through routes config, see if we can respond immediately (redirection or proxy_pass)
     - if not, instal builders, run build, find if there's a build output correspond to the request
       - if found a `Lambda`, invoke it and respond with result.
       - if found a `FileFsRef`, serve the file.
       - if found `undefined`, respond 404.

## TODO

- [x] Fix lambda invocation (blocking)
- [x] Import `@zeit/fun` from npm package
- [x] Env vars defined in `now.json` need to be present in lambda functions
- [ ] Resolve secret env vars for local dev
- [x] <strike>Fix `ncc` bundling issues to make `@now/node` work properly</strike> (standing issues are being worked around by the `build.sh` script in this PR)
- [ ] Add tests
- [ ] Refine logging & wording
- [x] `index.*` matching not yet working (@MAPESO)
- [x] `routes` are buggy (@DiegoRBaquero)
- [ ] `@now/build-utils`
    - [ ] Upgrade to latest and for TypeScript support
    - [ ] Silent mode (mute logs while building)
    - [x] Output `ncc`, `user` to temp dir instead of working directory (for `@now/node` + `@now/node-server`) https://github.com/zeit/now-builders/pull/245
- [ ] Test official builders:
  - [x] `@now/bash`
  - [x] `@now/node`
  - [x] `@now/node-server`
  - [x] `@now/go`
  - [x] `@now/python`
  - [x] `@now/next`
  - [ ] `@now/php`
  - [ ] `@now/rust`
  - [x] `@now/md`

amio committed:
[WIP] now-dev

amio committed:
try invoking lambda

amio committed:
add error status to dev-server

amio committed:
add basic routing

amio committed:
add entry for custom @now/static builder

amio committed:
honor .gitignore & .nowignore while collecting files

amio committed:
add route for single static file

amio committed:
add test fixtures for now-dev (nodejs)

amio committed:
fix miss typed @now/statics

amio committed:
import types from @now/build-utils

amio committed:
fix typo @now/statics -> @now/static

amio committed:
revamped router

amio committed:
Merge remote-tracking branch 'zeit/canary' into now-dev

amio committed:
typescript-eslint-parser -> @typescript-eslint/parser

Eliminate error while running pre-commit hooks:

  SUPPORTED TYPESCRIPT VERSIONS: ~3.1.1
  YOUR TYPESCRIPT VERSION: 3.2.2

amio committed:
send files from local @now/static builder

amio committed:
use serve-handler when no now.json

amio committed:
remove http-proxy

amio committed:
add type def for serve-handler

amio committed:
code cleanup

amio committed:
move type declarations to types.ts

amio committed:
dont rely on `this.cwd` while running builder

amio committed:
more type definitions

amio committed:
fix invoking error

amio committed:
clean code

amio committed:
better debug logging

amio committed:
refined request handler logic

amio committed:
handle /favicon.ico with serve-handler

amio committed:
make builder-cache self contained

amio committed:
move builder works to dev-builder.ts

amio committed:
add proxy_pass handler

amio committed:
support named groups in routes

amio committed:
fix creating cache dir error

amio committed:
add "--nodejs-preview" solely for nodejs

amio committed:
clean require.cache after required user js

amio committed:
fix router & header issue in nodejs preview

amio committed:
[wip] add tests for dev-router

amio committed:
Merge remote-tracking branch 'zeit/canary' into now-dev

amio committed:
add tests for dev-router

amio committed:
lookup "/index.*" for indexes

amio committed:
refine logging

amio committed:
Merge remote-tracking branch 'zeit/canary' into now-dev

TooTallNate committed:
[WIP] `now dev` progress point

Actually invoking the `@zeit/fun` module at this point.

Still lots of work to be done…

TooTallNate committed:
Update `@zeit/fun` to v0.0.4

TooTallNate committed:
Remove `--nodejs-preview`

To be replaced by `@zeit/fun`.

TooTallNate committed:
Remove `decache` and `import-fresh`

Leftover from the `--nodejs-preview`.

TooTallNate committed:
Remove "declaration: true" from `tsconfig.json`

Otherwise the ncc `dist` dir ends up full of generated `.d.ts` files.

TooTallNate committed:
Remove `ignore` dep, always use `@zeit/dockerignore`

TooTallNate committed:
Use tarball URL for `@zeit/fun`

Because `@zeit/fun` is not yet public on npm.

amio commented on package.json:
```diff
@@ -136,7 +136,7 @@
     "@typescript-eslint/parser": "1.1.0",
     "@zeit/dockerignore": "0.0.4",
-    "@zeit/fun": "0.0.4",
+    "@zeit/fun": "http://zeit-fun-0.0.5.zeit.sh",
```
👍, got an error while running `yarn install`:

```
error http://zeit-fun-0.0.5.zeit.sh: Integrity check failed for "@zeit/fun" (computed integrity doesn't match our records, got "sha1-HkLvdBQ3A8MQc519II8IfDwoiWw=")
```

TooTallNate replied:
Sorry about that @amio. Should be fixed by 7e668cea6a65be7bcf695b9c306520636d7c6a4e.

TooTallNate committed:
Regenerate `yarn.lock` to fix bad integrity check on fun tarball

TooTallNate committed:
Upgrade to `@zeit/fun` v0.0.6

With re-written unzipping logic utilizing `yauzl`.

TooTallNate committed:
Add lambda function handling logic for the HTTP response

TooTallNate committed:
Match the lambda invoke shape that Now's proxy provides

TooTallNate committed:
Move deps to devDependencies

TooTallNate committed:
Update `@zeit/fun` to v0.0.8

TooTallNate committed:
Update `@zeit/ncc` to v0.15.0

TooTallNate committed:
Attempt to fix unit test for `wait()` helper function

TooTallNate committed:
Add `build.sh` script to work around ncc issues

TooTallNate committed:
Fix `build.sh` script for BSD's chmod

No `--changes` flag.

TooTallNate committed:
Update `@zeit/fun` to v0.0.11

TooTallNate committed:
Update `@zeit/fun` to v0.0.12

TooTallNate committed:
Tweak some comments

TooTallNate committed:
Update `@zeit/fun` to v0.0.13

TooTallNate committed:
Add `now.json` env vars to lambda functions

pie6k commented:
It looks awesome! 

I've got one question - what would happen if I modify some file (eg. Lambda function) during dev mode?

Would I have to restart dev mode to have it refreshed or it would happen automatically? 

It would be very useful when you have to iterate quickly (eg. debugging)

DiegoRBaquero commented:
@pie6k Not yet, but hot reloading is something that is def going to be added.

TooTallNate committed:
Finish removing `--nodejs-preview` flag

TooTallNate committed:
Style tweaks

TooTallNate committed:
Use `npm-package-arg` to parse the "use" builder

TooTallNate committed:
Styling

TooTallNate committed:
Use the parsed `pathname` when matching routes

The querystring portion needs to be removed to match properly.

TooTallNate committed:
Update `@zeit/fun` to v0.1.1

TooTallNate committed:
Merge remote-tracking branch 'origin/canary' into now-dev

TooTallNate committed:
Support `-d` for debug flag

It was already listed as such in the `--help` output.

TooTallNate committed:
List the `dist/runtimes` dir as pkg "assets"

TooTallNate committed:
Update `@now/build-utils` to v0.4.37-canary.0

TooTallNate committed:
Update `@zeit/fun` to v0.2.0

TooTallNate committed:
Merge branch 'canary' into now-dev

TooTallNate committed:
Update `@zeit/fun` to v0.2.1

aequasi commented:
Not sure if its the right place to ask for it, but it would be handy to have the capability to also run a proxy with `now dev`, so you could run https://dev.mydomain.com

Something like: https://www.npmjs.com/package/dev-proxy

TooTallNate commented:
@aequasi This is on the roadmap as well, but probably won't be part of the initial release.

TooTallNate committed:
Rewrite the `req.url` to fix serving FileFsRef assets

TooTallNate committed:
Disable re-building, build all `builds` upon server bootup

TooTallNate committed:
404 is not a redirect

amio committed:
warning on more than one argument

TooTallNate committed:
Fix `index` without a file extension matching

TooTallNate committed:
Update `@zeit/fun` to v0.2.2

TooTallNate committed:
Use `async-listen`

TooTallNate committed:
Add shutdown cleanup logic

TooTallNate committed:
Rename `buildLambdas()` to `executeBuilds()`

TooTallNate committed:
Rename `lambda` prop to `fn`

TooTallNate committed:
Misc tweaks

TooTallNate committed:
Pass `isDev: true` to the builder `build()` function

TooTallNate committed:
Use `Object.assign()` instead of object spread

TooTallNate committed:
Set `res.statusCode` instead of `writeHead()`

TooTallNate committed:
Create Lambda functions during build-time

This way, the lambda instance is re-used for subsequent HTTP requests

amio commented on src/commands/dev/lib/dev-builder.ts:
```diff
@@ -74,7 +72,7 @@ async function executeBuilds(
           isDev: true
         });
-        results = { ...results, ...output };
+        Object.assign(results, output);
```
Why prefer `Object.assign` over object spread? 🤔

TooTallNate replied:
It's a very minor optimization, but in this case the `results` object is not being re-created upon each pass of the loop.

TooTallNate committed:
Wait for cleanup operations to complete in `stop()`

TooTallNate committed:
Destructure `path` for join()

TooTallNate committed:
Add some missing types

TooTallNate committed:
Store the `buildConfig` on the assets for reverse lookup

TooTallNate committed:
Remove `console.log()`

TooTallNate committed:
Attempt to make globbing faster and ignore files more accurately

TooTallNate committed:
Rewrite `installBuilders()` to be idempotent

So that npm is not executed for canary tags, for example.

This is done by comparing the contents of the package.json file rather
then the `node_modules` directory.

amio commented on src/commands/dev/lib/dev-server.ts:
```diff
@@ -79,11 +80,27 @@ export default class DevServer {
    */
   async start(port: number = 3000): Promise<void> {
-    const nowJson = readLocalConfig(this.cwd);
+    const nowJson = await this.getNowJson();
```
@TooTallNate Did you create this custom `getNowJson` to avoid those `process.exit(1)` in [`readLocalConfig`](https://github.com/zeit/now-cli/blob/5a111186f4a47e0b27d4ee1de9c365f9da6baf11/src/util/config/files.ts#L71-L98)?

TooTallNate replied:
The plan is that we can optimize this function to cache the result, and also add validation (i.e. for `version: 2`). It will then be invalidated with a watcher / periodic shasum / whatever mechanism we decide on to reduce the number of times the file is being read + parsed.

TooTallNate committed:
Use `ignore` instead of `@zeit/dockerignore` for glob-gitignore

It seems that the object produced by `@zeit/dockerignore` is not
compatible with the `glob-gitignore` module, so use the vanilla
`ignore` module instead for now as a workaround.

TooTallNate committed:
Implement a `GlobIgnore` class to replace `glob-gitignore`

This allows us to continue using `@zeit/dockerignore` for proper
consistency.

Closes #1899.

TooTallNate committed:
Prettier

TooTallNate committed:
Skip installing built-in builders (i.e. `@now/static`)

TooTallNate committed:
Remove `console.error()` calls

TooTallNate committed:
Use a deterministic `workPath` directory

So that the same one is re-used in-between subsequent boots of `now dev`

TooTallNate committed:
Add `now.json` validation logic

TooTallNate committed:
Fix serving static assets outside the project working directory

This fixes, for example, `@now/mdx-deck` which outputs static files
from its build process.

TooTallNate committed:
Add comment

TooTallNate committed:
Validate the `now.json` file before launching the dev server

TooTallNate committed:
Respect `maxLambdaSize` config

Closes #1950.

TooTallNate committed:
Add `@types/bytes` dependency

styfle commented on src/commands/dev/lib/types.ts:
```diff
@@ -0,0 +1,129 @@
+
+export interface Builder {
+  config?: {
+    maxLambdaSize: string | number;
```
```suggestion
    maxLambdaSize?: string | number;
```

I think this config is optional so you want to make sure it exists before reading it.

MAPESO replied:
What happens if it is not optional ?

MAPESO replied:
bwt **maxLambdaSize** is having an error, apparently it is not requested now-dev when dependencies are installed that require more than **5mb** 🤔

styfle committed:
Make `maxLambdaSize` be optional

Co-Authored-By: TooTallNate <n@n8.io>

TooTallNate committed:
WIP force reload in browser = rebuild asset behavior

TooTallNate committed:
Resolve the `asset` again after re-build

TooTallNate committed:
Remove logging

TooTallNate committed:
Update `@zeit/fun` to v0.3.0

Mathspy commented:
Small reminder: `@now/rust` was released after this PR was created so it's not on the original post's list. I wouldn't mind being the one to test it.

And thank you for all the great work :heart:

TooTallNate committed:
Update `@now/build-utils` to v0.4.37-canary.1

Sets `npm_config_target` to the proper Node.js version for native
modules.

TooTallNate committed:
Update `@zeit/fun` to v0.3.1

TooTallNate committed:
Fix "hard refresh to reload" after the first rebuild

The `buildConfig` and `buildEntry` proper were not being set on the
rebuilt assets, causing the rebuild to only work the first time.

Now fixed.

TooTallNate committed:
Merge remote-tracking branch 'origin/canary' into now-dev

markozxuu committed:
Add logic for the `now dev cache clean` command (#1977)

* Added the cache-clean command

* censer code :)

TooTallNate committed:
Fix non-route match `dest` param

This makes the `go-image-to-ascii` work, since the query param URL gets
properly routed to the `/` route.

TooTallNate committed:
Add support for `FileBlob` outputs

Fixes `@now/optipng`, `@now/html-minifier` and `@now/md` builders.

TooTallNate committed:
Remove `console.error()` call

styfle commented on package.json:
```diff
@@ -116,32 +119,51 @@
+    "@zeit/fun": "https://zeit-fun-0-3-1.zeit.sh",
     "@zeit/git-hooks": "0.1.4",
-    "@zeit/ncc": "0.8.1",
+    "@zeit/ncc": "0.15.0",
```
```suggestion
    "@zeit/ncc": "0.16.1",
```

MAPESO replied:
Good catch @styfle  👌

TooTallNate committed:
Add `now dev cache clean` command

MAPESO commented on src/commands/dev/lib/builder-cache.ts:
```diff
@@ -56,10 +59,22 @@ export async function prepareBuilderDir() {
-    const deleteCache = await cacheDirPromise;
-    await remove(deleteCache);
+    const cacheDir = await cacheDirPromise;
+    output.log(chalk`{magenta Deleting} ${cacheDir}`);
```
well thought, with that we are indicating to the user that the chache is being erased 👌

TooTallNate committed:
Use the `assetKey` to look up the Content-Type on FileBlobs

Instead of the `fsPath` of the entrypoint file.

This fixes `@now/md` to return `text/html` Content-Type instead of
`text/markdown` which does not render properly in the web browser.

styfle committed:
Update `@zeit/ncc` to v0.16.1

Co-Authored-By: TooTallNate <n@n8.io>

TooTallNate committed:
Regenerate `yarn.lock` file

styfle commented on src/commands/dev/lib/builder-cache.ts:
```diff
@@ -0,0 +1,150 @@
+export async function cleanCacheDir(output: Output): Promise<void> {
+  try {
+    const cacheDir = await cacheDirPromise;
+    output.log(chalk`{magenta Deleting} ${cacheDir}`);
```
```suggestion
    output.log(chalk.magenta(`Deleting ${cacheDir}`));
```

This looks like a typo to me 🤔

TooTallNate replied:
It's not a typo, `chalk` supports this syntax. Also note that your proposal colors the cacheDir magenta, whereas the original code does not.

styfle commented on src/commands/dev/lib/builder-cache.ts:
```diff
@@ -0,0 +1,150 @@
+
+  try {
+    await remove(funCacheDir);
+    output.log(chalk`{magenta Deleting} ${funCacheDir}`);
```
```suggestion
    output.log(chalk.magenta(`Deleting ${funCacheDir}`));
```

This looks like another typo

TooTallNate committed:
Remove leading `/` on build `src` specifications

TooTallNate committed:
Use `rsync` to copy over all of `fun`'s runtime files

sushruth commented:
I am waiting for this so much!
![](https://66.media.tumblr.com/tumblr_meirg8LP2B1r1mr1po1_500.gif)

TooTallNate committed:
Update `@zeit/fun` to v0.5.0

TooTallNate committed:
Update `@zeit/fun` to v0.5.1

TooTallNate committed:
try/catch the lambda invoke and send 500 upon error

styfle commented on @types/@now_build-utils/index.d.ts:
```diff
@@ -0,0 +1,107 @@
+  export default Lambda;
+}
+
+declare module '@now/build-utils' {
```
You can probably remove these now

TooTallNate replied:
Done 810b951f83b3ac9e3ab425a1f4461b1cc9d414d6

TooTallNate committed:
Add `rsync` to CircleCI build

ecwyne commented:
Gave `~/Downloads/now-macos dev` a try thing morning (thanks for the detailed instructions!) and got this error `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`

MAPESO commented:
> Gave `~/Downloads/now-macos dev` a try thing morning (thanks for the detailed instructions!) and got this error `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`
> 
> My now.json has 3 builds:
> 
>  ```
> "builds": [
>         {
>             "src": "www/package.json",
>             "use": "@now/next"
>         },
>         {
>             "src": "api/index.ts",
>             "use": "@now/node-server@canary"
>         },
>         {
>             "src": "api/methods/*.ts",
>             "use": "@now/node@canary"
>         }
>     ],
> ```
> Doing some research I found [this thread](https://github.com/npm/npm/issues/12238) very useful and was directed to [this package](https://www.npmjs.com/package/increase-memory-limit).
> 
> Hope that's helpful!
> 
> full error output

Hi @ecwyne 👋 😸 Thanks for giving you the opportunity to try `now-dev` : ]

currently all the tests we are doing of the constructors in `now-dev` use the stable version, with this it should work.  If the problem persists, let me know 🙏 

```json
"builds": [
        {
            "src": "www/package.json",
            "use": "@now/next"
        },
        {
            "src": "api/index.ts",
            "use": "@now/node-server"
        },
        {
            "src": "api/methods/*.ts",
            "use": "@now/node"
        }
    ]
```

ecwyne commented:
Thanks @MAPESO, just gave it a try with `@now/node` and `@now/node-server` instead of the canary versions and I still get the same error `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`

MAPESO commented:
> Thanks @MAPESO, just gave it a try with `@now/node` and `@now/node-server` instead of the canary versions and I still get the same error `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`

try running the following command `DEBUG=*fun* ~/Downloads/now-macos dev --debug`

MAPESO commented:
@ecwyne   I'm sure that should solve the problem 😄

TooTallNate commented:
@ecwyne Thanks for trying it out. Would you be able to share some code to repro that on my end? Thanks in advance!

TooTallNate committed:
Merge remote-tracking branch 'origin/canary' into now-dev

TooTallNate committed:
Convert `errors.js` to TypeScript to fix compilation errors

TooTallNate committed:
Also accept `Cache-Control: no-cache` for rebuilding

TooTallNate committed:
Prettier

TooTallNate committed:
Update `@now/build-utils` to v0.4.38-canary.1

markozxuu committed:
Update `@zeit/ncc` to v0.17.0 (#2017)

martonlanga commented:
I'm getting Build errors for not resolved routes on a next.js project

`{ Error: Can't resolve '../components/User/Signup' in '/Users/martonlanga/Library/Caches/co.zeit.now/dev/workPaths/24280926/www/pages'`

It works when deployed or running next.js dev server

MAPESO commented:
@martonlanga Hi, thanks for trying `now-dev` it would be great if we could take a look at the code to go deeper into the problem 🙏 ❤️ 

the internal tests that we carry out this type of projects should work normally.

martonlanga commented:
That was weird.

Checked the Cache at `../Library/Caches/co.zeit.now/dev/workPaths/24280926/www/components` and the `User` folder was missing. Renamed it to `Logger` and now the error is gone ¯\_(ツ)_/¯

martonlanga commented:
Had to install `yarn` as well to make it work. Might be useful to add that to the install guide.

Does env vars resolvement work already? It seems like it is for me

Ehbraheem commented:
Anyone tried `now-dev` on windows?

raulg commented:
Getting `Error: Unable to encode undefined of type undefined` when trying to navigate directly to a page defined in the routes, `/page?uid=$uid` in a project integrating Prismic.
Issue seems to be when getInitialProps is called and tries to fetch using the Prismic API
```
const API = await Prismic.getApi(apiEndpoint, { req, accessToken });
const document = await API.getByUID('page', uid);
```

Works without any issues when deployed in Now.

homerjam commented:
I just tested with `@now/php` and I get the following..
```
 >>> GET /
{ Error: spawn ./php-fpm ENOENT
    at _errnoException (util.js:1022:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:190:19)
    at onErrorNT (internal/child_process.js:372:16)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
  code: 'ENOENT',
  errno: 'ENOENT',
  syscall: 'spawn ./php-fpm',
  path: './php-fpm',
  spawnargs: [ '-c', 'php.ini', '--fpm-config', 'php-fpm.ini', '--nodaemonize' ] }
```
Looking forward to being able to use this!

MAPESO commented:
> Checked the Cache at `../Library/Caches/co.zeit.now/dev/workPaths/24280926/www/components` and the `User` folder was missing. Renamed it to `Logger` and now the error is gone ¯_(ツ)_/¯

We developed a `now dev cache clean` command that avoids these problems related to the cache. Hopefully, it helps to avoid this kind of mistakes 🙏

MAPESO commented:
> Had to install `yarn` as well to make it work. Might be useful to add that to the install guide.
> 
> Also, does env vars resolvement work already? It seems like it is for me
> And is it possible that the `NODE_ENV` is set to production?

We will add the support before launching `now-dev` : ]

MAPESO commented:
> I just tested with `@now/php` and I get the following..
> 
> ```
>  >>> GET /
> { Error: spawn ./php-fpm ENOENT
>     at _errnoException (util.js:1022:11)
>     at Process.ChildProcess._handle.onexit (internal/child_process.js:190:19)
>     at onErrorNT (internal/child_process.js:372:16)
>     at _combinedTickCallback (internal/process/next_tick.js:138:11)
>     at process._tickCallback (internal/process/next_tick.js:180:9)
>   code: 'ENOENT',
>   errno: 'ENOENT',
>   syscall: 'spawn ./php-fpm',
>   path: './php-fpm',
>   spawnargs: [ '-c', 'php.ini', '--fpm-config', 'php-fpm.ini', '--nodaemonize' ] }
> ```
> Looking forward to being able to use this!

We are working on support for `php` 🔜  :D

MAPESO commented:
> Anyone tried `now-dev` on windows?

In case you want to use windows I recommend using the bash of "Ubuntu" has many advantages related to permissions etc. 

run the following commands:
```bash
chmod +x ~/Downloads/now-win
~/Downloads/now-win dev
```

MAPESO commented on yarn.lock:
```diff
@@ -521,10 +521,10 @@
-  version "0.16.1"
-  resolved "https://registry.yarnpkg.com/@zeit/ncc/-/ncc-0.16.1.tgz#c12257d2d919fc47447944f5021b33e5997a84fb"
-  integrity sha512-lHScAn2rl2OjBzpzFZlv9YCq9ESvcBz51NWxXV7D1nA02jceUoIqlK8YDQSmA5wa5IqL+UzFyxSpq6ZSyVSAHw==
+"@zeit/ncc@0.17.0":
```
So sad. This does not solve the problem of `firestore` with `now-dev` 🙍 💔

MAPESO replied:
@TooTallNate  If this works  in `ncc` the problem must be `now-dev`

TooTallNate committed:
Merge remote-tracking branch 'origin/canary' into now-dev

TooTallNate committed:
Install `@zeit/fun` from npm

juliangruber commented on build.sh:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+set -euo pipefail
```
this doesn't use pipes so I assume it's a precaution?

TooTallNate replied:
Ya this is just my "standard" Bash configuration. Also see https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/ (`-x` is too noisey for me though).

juliangruber commented on src/commands/dev/lib/builder-cache.ts:
```diff
@@ -0,0 +1,150 @@
+      code: 'NO_BUILDER_CACHE_DIR',
+      message: 'Could not find cache directory for now-builders.',
+      meta: {}
+    });
```
Errors are usually added to https://github.com/zeit/now-cli/blob/canary/src/util/errors-ts.ts

juliangruber commented on src/commands/dev/lib/builder-cache.ts:
```diff
@@ -0,0 +1,150 @@
+      code: 'NO_CLEAN_CACHE',
+      message: 'Error clearing the cache',
+      meta: {}
+    });
```
Same about Error classes

MAPESO replied:
@juliangruber I wrote this error message, is there something I should modify?

juliangruber replied:
See https://github.com/zeit/now-cli/pull/1883#discussion_r269275009

juliangruber commented on src/commands/dev/lib/builder-cache.ts:
```diff
@@ -0,0 +1,150 @@
+      code: 'NO_CLEAN_CACHE',
+      message: 'Error clearing the cache',
+      meta: {}
+    });
```
Error classes

juliangruber commented on src/commands/dev/lib/dev-builder.ts:
```diff
@@ -0,0 +1,296 @@
+  devServer: DevServer,
+  asset: BuilderOutput
+): Promise<void> {
+  //console.error('rebuilding', { asset });
```
comment
```suggestion
```

juliangruber commented on src/commands/dev/lib/dev-builder.ts:
```diff
@@ -0,0 +1,296 @@
+          ).toLowerCase()}) exceeds the configured limit (${bytes(
+            maxLambdaBytes
+          ).toLowerCase()}). You may increase this by supplying \`maxLambdaSize\` to the build \`config\``
+        );
```
Error class

juliangruber commented on src/commands/dev/lib/dev-builder.ts:
```diff
@@ -0,0 +1,296 @@
+  }
+
+  await Promise.all(
+    Object.entries(output).map(async e => {
```
Is `e` an event? Is that a known convention or could we use a more descriptive name?

juliangruber commented on src/commands/dev/lib/dev-builder.ts:
```diff
@@ -0,0 +1,296 @@
+
+  if (!nowJson.builds) {
+    return;
+  }
```
should we move this to the function start?

TooTallNate replied:
Ya good call

juliangruber commented on src/commands/dev/lib/dev-builder.ts:
```diff
@@ -0,0 +1,296 @@
+                ).toLowerCase()}) exceeds the configured limit (${bytes(
+                  maxLambdaBytes
+                ).toLowerCase()}). You may increase this by supplying \`maxLambdaSize\` to the build \`config\``
+              );
```
Error class

juliangruber commented on src/commands/dev/lib/dev-builder.ts:
```diff
@@ -0,0 +1,296 @@
+        message: `Failed building ${chalk.bold(build.src)} with ${build.use}`,
+        meta: err.stack
+      });
+      */
```
comment

juliangruber committed:
Remove `console.error()` call

Co-Authored-By: TooTallNate <n@n8.io>

TooTallNate committed:
Create `NowError` subclasses

TooTallNate committed:
Add `LambdaSizeExceededError`

TooTallNate committed:
Return early if there's no builds

TooTallNate commented:
Thank you to everyone who gave your input. We're going to merge this PR now as the initial iteration of `now dev`. There's still lots of improvements planned, but we want to get this into your hands now and continue iterating from there.

If you have any issues with `now dev` going forward, please create a new issue and add the `now dev` label to the issue.

Thank you all!

giovannibonin commented:
Is there a way to try this by installing with NPM? I just installed version
14.3.0-canary.17 & I get " Error! The specified file or directory "dev"
does not exist." when running "now dev" in a project folder. Am I missing
something?

Thanks
Giovanni

On Tue, Mar 26, 2019 at 4:47 PM Nathan Rajlich <notifications@github.com>
wrote:

> Merged #1883 <https://github.com/zeit/now-cli/pull/1883> into canary.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zeit/now-cli/pull/1883#event-2231502623>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/Ati53AWrZgkmr8Qm1ZO-DKx746Jra-AUks5varGhgaJpZM4askd0>
> .
>

TooTallNate commented:
@giovannibonin It's not yet published, but I'm working on that. Waiting for CircleCI to be green.

TooTallNate commented:
Ok it's published in `14.3.0-canary.18` 👍

pmuellr commented:
just a quick note - I had `yarn` installed globally at 1.7.0 (I guess very old!) and `now dev` ended up failing with a message:

```
error An unexpected error occurred: "ENOENT: no such file or directory, scandir '/Users/pmuellr/Library/Caches/Yarn/v1/npm-@zeit'".
```

Upgraded yarn to 1.15.2, and `now dev` works a treat!

\o/

brettneese commented:
I can beat @pmuellr. My Yarn version was v1.3.2 and I ran into the same problem . Thanks for the heads-up Patrick!

agstover commented:
I think there's an issue with now dev not pulling in build.env as defined in the now.json file. I created an issue here: https://github.com/zeit/now-builders/issues/333

Basically the same code run with 'now' accesses the build env variables in next.config.js, but the code run with 'now dev' returns undefined for those variables.

TooTallNate commented:
@agstover Thanks for the report. Could you create a new issue for that please?

khrome83 commented:
A few questions - 

1. Should `now dev` be rebuilding everything and reinstalling from yarn with each build? When i change a string in a `index.js` file that makes a hello world lambda, it redownloads with Yarn, and rebuilds Nuxt Generate which is in a different folder...

2. How to kill `now dev`. My only choice is to use `ctrl+z`. This leaves the process running. I had to do this because it did not seem sometimes for some files it rebuild. For example changing a .vue file. Upon killing, the port used is consumed. Running `ps` I can see the used process id, running `kill #####` does not appear to actually kill the process, and I had to restart my machine. Running on a 2017 Macbook Pro, latest OS.

nathankallman commented:
@khrome83, `ctrl + c` should work for killing the server. This is standard on *nix like command lines to send the `SIGINT` signal to the current process. (Where `ctrl + z` sends `SIGSTOP`)

https://superuser.com/questions/262942/whats-different-between-ctrlz-and-ctrlc-in-unix-command-line

khrome83 commented:
@nathankallman - it was not doing anything... That was the first thing I tried. It just ignored it.

nathankallman commented:
@khrome83, that is probably a bug then. `ctrl + c` worked for me on `now-cli` `15.0.4`, but so far I've only used markdown and static builders...

Have you tried `kill -9 <pid>` or `sudo kill -9 <pid>`?

khrome83 commented:
@nathankallman - I will give that a shot when I try tonight, and report back. Good idea. My Unix is rustier than I care to admit.

kemptone commented:
I'm new to now and zeit, but have been spending some time in the last two days/nights trying to get ```now dev``` working for testing locally. Based on the now-examples, it seems most are none functional with ```now dev```.  Are there some specific now 2.0 examples this is being tested against that might serve as an example to start with? If so, where do I see them?

The specific kind of configuration I want to try is quite simple. A static folder, with some lambda node functions. To be even more specific, I want to export a next.js site to a ```dist/``` folder locally, and then serve that statically, but also be able to serve some lambdas. I've tried nearly all of the example sites and have yet to find that specific combination.

Regardless, when I try this configuration, via now dev:

```{
  "version": 2,
  "name": "writersmugs",
  "builds": [
    { "src": "/dist/**", "use": "@now/static" }
  ],
  "routes" : [
    { "src" : "/(.*)", "use" : "/dist/$1" }
  ]
}```

this is the output:
```
writersmugs git:(master) ✗ now dev
> NOTE: Requested port 3000 is already in use
> Ready! Development server running at http://localhost:3001
> GET /about
```

but the browser gives me:
404: The page could not be found
Code: FILE_NOT_FOUND

williamli commented:
@kemptone `now-dev` is currently still in beta, we will be releasing more docs, and guides shortly. Please stay tuned to our updates on our blog at https://zeit.co/blog.

khrome83 commented:
@nathankallman - I think it was an issue with `vs code` and the recent terminal issue they had caused it to be unresponsive. I updated to the March release of the IDE and it seems to resolve the issue. =)

kemptone commented:
@williamli thanks for all you guys do. I figured out my issue thanks to the -d flag.
